tools/restatectl/src/commands/replicated_loglet/storage_state.rs

```diff
@@ -124,165 +69,168 @@ async fn set_storage_state(connection: &ConnectionInfo, opts: &SetOpts) -> anyho
+                (StorageState::ReadWrite, _) => false,
+                (StorageState::DataLoss, StorageState::ReadWrite) => {
+                    // that's only possible if we are okay with either:
+                    // - declaring permanent data-loss of data on this node
+                    // - the node transitioned into data-loss, but we didn't actually lose data.
+                    false
+                }
+                (StorageState::DataLoss, StorageState::Provisioning) => false,
+                (StorageState::DataLoss, StorageState::ReadOnly) => {
+                    // silent under-replication
+                    false
+                }
+                (StorageState::DataLoss, StorageState::Disabled) => {
+                    // only if data were trimmed
+                    false
+                }
+                (_, _) => false,
+            };
+
+            if !safe {
+                return Err(anyhow::anyhow!(
+                    "This node is currently in `{current_state}` storage-state. Transitioning into `{}` is unsafe.",
+                    opts.storage_state
+                ));
+            }
         }
+    }
 
-        let safe = match (current_state, opts.storage_state) {
-            (from, to) if from == to => {
-                c_println!(
-                    "Node {} storage-state was already {}",
-                    opts.node_id,
-                    current_state
+    // run safety checks
+    if !opts.force && !opts.storage_state.can_write_to() {
+        let logs = connection.get_logs().await?;
+        let provider = &logs.configuration().default_provider;
+        if let ProviderKind::Replicated = provider.kind() {
+            // any log-id would do.
+            let options = NodeSetSelectorOptions::new(0)
+                .with_target_size(provider.target_nodeset_size().unwrap());
+            let replication_property = provider.replication().unwrap();
+            let nodeset = NodeSetSelector::select(
+                &nodes_config,
+                replication_property,
+                |node_id, config| {
+                    // skip our selection of nodes from being candidate
+                    if opts.nodes.contains(&node_id) {
+                        false
+                    } else {
+                        logserver_candidate_filter(node_id, config)
+                    }
+                },
+                |_, config| {
+                    matches!(
+                        config.log_server_config.storage_state,
+                        StorageState::ReadWrite
+                    )
+                },
+                options,
+            )
+            .context("Safety-check failed: lost the ability to generate nodesets")?;
+
+            let mut checker = NodeSetChecker::new(&nodeset, &nodes_config, replication_property);
+            checker.fill_with(true);
+            if !checker.check_write_quorum(|attr| *attr) {
```

Comment:
In which case will the `NodeSetSelector` return a nodeset that does not fulfill write quorum if all nodes are marked as `true`?

Reply:
For instance, if the total number of nodes > num_copies, but it cannot form a write quorum according to locality rules (10 nodes, in the same region, but replication is {region: 3}) for example.